core/account/reserve.go

```diff
@@ -233,41 +243,74 @@ func (re *reserver) source(src source) *sourceReserver {
 	sr = &sourceReserver{
 		db:       re.db,
 		src:      src,
+		validFn:  re.checkUTXO,
+		cached:   make(map[bc.Outpoint]*utxo),
 		reserved: make(map[bc.Outpoint]uint64),
 	}
 	re.sources[src] = sr
 	return sr
 }
 
 type sourceReserver struct {
-	db    pg.DB
-	src   source
-	group singleflight.Group
+	db      pg.DB
+	src     source
+	validFn func(u *utxo) bool
 
 	mu       sync.Mutex
+	cached   map[bc.Outpoint]*utxo
 	reserved map[bc.Outpoint]uint64
 }
 
-func (sr *sourceReserver) findMatchingUTXOs(ctx context.Context) ([]*utxo, error) {
-	srcID := fmt.Sprintf("%s-%s", sr.src.AssetID, sr.src.AccountID)
-	untypedUTXOs, err := sr.group.Do(srcID, func() (interface{}, error) {
-		return findMatchingUTXOs(ctx, sr.db, sr.src)
-	})
-	return untypedUTXOs.([]*utxo), err
-}
-
 func (sr *sourceReserver) reserve(ctx context.Context, rid uint64, amount uint64) ([]*utxo, uint64, error) {
 	var reserved, unavailable uint64
 	var reservedUTXOs []*utxo
 
+	// First try to reserve using only cached UTXOs.
+	sr.mu.Lock()
+	for o, u := range sr.cached {
+		// If the UTXO is already reserved, skip it.
+		if _, ok := sr.reserved[u.Outpoint]; ok {
+			continue
+		}
+		// Cached utxos aren't guaranteed to still be valid; they may
+		// have been spent. Verify that that the outputs are still in
+		// the state tree.
+		if !sr.validFn(u) {
+			delete(sr.cached, o)
+			continue
+		}
+
+		reserved += u.Amount
+		reservedUTXOs = append(reservedUTXOs, u)
+		if reserved >= amount {
+			break
+		}
+	}
+	if reserved >= amount {
+		// We've found enough to satisfy the request.
+		for _, u := range reservedUTXOs {
+			sr.reserved[u.Outpoint] = rid
+			delete(sr.cached, u.Outpoint)
+		}
+		sr.mu.Unlock()
+		return reservedUTXOs, reserved, nil
+	}
+	sr.mu.Unlock()
+	reserved = 0
+	reservedUTXOs = nil
+
 	// Find the set of UTXOs that match this source.
-	utxos, err := sr.findMatchingUTXOs(ctx)
+	utxos, err := findMatchingUTXOs(ctx, sr.db, sr.src)
```

Comment:
If we didn't delete reserved utxos, we could add a cursor here, and only load new utxos. We could even only do that if the processor pin height is greater than the last cursor height.